File: apps/mobile/lib/features/plans/plan_steps.dart
/// 计划阅读 Step：最小跳转单元。
library;

class PlanStep {
  const PlanStep({
    required this.id,
    required this.label,
    required this.bookId,
    required this.chapterStart,
    required this.chapterEnd,
  });

  final String id;
  final String label;
  final String bookId;
  final int chapterStart;
  final int chapterEnd;
}

List<PlanStep> stepsForReadingRows(List<Map<String, dynamic>> rows, int day) {
  return rows
      .where((r) => (r['day'] as int?) == day)
      .map((r) {
        final book = (r['book'] ?? '') as String;
        final cs = (r['chapter_start'] ?? r['chapterStart'] ?? 1) as int;
        final ce = (r['chapter_end'] ?? r['chapterEnd'] ?? cs) as int;
        final name = (r['book_name'] ?? r['bookName'] ?? book) as String;
        final title = (r['title'] ?? '') as String;
        return PlanStep(
          id: '$book.$cs-$ce',
          label: title.isNotEmpty ? title : _formatRange(name, cs, ce),
          bookId: book.toUpperCase(),
          chapterStart: cs,
          chapterEnd: ce,
        );
      })
      .toList();
}

List<PlanStep> stepsFromRefs(List<String> refs, {String? titleHint}) {
  if (refs.isEmpty) return [];
  final parsed = refs.map(_parseRef).toList();
  final steps = <PlanStep>[];
  var start = 0;
  for (var i = 1; i <= parsed.length; i++) {
    final prev = parsed[i - 1];
    final curr = i < parsed.length ? parsed[i] : null;
    final breakGroup = curr == null ||
        curr.$1 != parsed[start].$1 ||
        curr.$2 != prev.$2 + 1;
    if (breakGroup) {
      final seg = parsed.sublist(start, i);
      final book = seg.first.$1;
      final cs = seg.first.$2;
      final ce = seg.last.$2;
      steps.add(PlanStep(
        id: '$book.$cs-$ce',
        label: _formatRange(book, cs, ce),
        bookId: book,
        chapterStart: cs,
        chapterEnd: ce,
      ));
      start = i;
    }
  }
  if (steps.length == 1 && titleHint != null && titleHint.isNotEmpty) {
    steps[0] = PlanStep(
      id: steps[0].id,
      label: titleHint,
      bookId: steps[0].bookId,
      chapterStart: steps[0].chapterStart,
      chapterEnd: steps[0].chapterEnd,
    );
  }
  return steps;
}

List<PlanStep> stepsForGeneratedDay(Map<String, dynamic> day) {
  final refs = ((day['refs'] ?? []) as List).cast<String>();
  return stepsFromRefs(refs, titleHint: day['title'] as String?);
}

({int done, int total}) sessionProgress(
    List<PlanStep> steps, List<String> stepsDone) {
  final done = steps.where((s) => stepsDone.contains(s.id)).length;
  return (done: done, total: steps.length);
}

bool allStepsDone(List<PlanStep> steps, List<String> stepsDone) =>
    steps.isNotEmpty && steps.every((s) => stepsDone.contains(s.id));

int stepForChapter(List<PlanStep> steps, String bookId, int chapter) {
  final bid = bookId.toUpperCase();
  return steps.indexWhere(
    (s) => s.bookId == bid && chapter >= s.chapterStart && chapter <= s.chapterEnd,
  );
}

bool isLastChapterOfStep(PlanStep step, int chapter) => chapter == step.chapterEnd;

PlanStep? nextIncompleteStep(List<PlanStep> steps, List<String> stepsDone) {
  for (final s in steps) {
    if (!stepsDone.contains(s.id)) return s;
  }
  return null;
}

PlanStep? pendingNextStep(
  List<PlanStep> steps,
  List<String> stepsDone,
  String bookId,
  int chapter,
) {
  final bid = bookId.toUpperCase();
  final step = steps.cast<PlanStep?>().firstWhere(
        (s) =>
            s!.bookId == bid &&
            chapter >= s.chapterStart &&
            chapter <= s.chapterEnd,
        orElse: () => null,
      );
  if (step == null || chapter != step.chapterEnd) return null;
  final next = nextIncompleteStep(steps, stepsDone);
  if (next == null || next.id == step.id) return null;
  return next;
}

(String bookId, int chapter) _parseRef(String ref) {
  final parts = ref.split('.');
  return (parts[0].toUpperCase(), int.tryParse(parts.length > 1 ? parts[1] : '1') ?? 1);
}

String _formatRange(String bookName, int start, int end) {
  if (start == end) return '$bookName $start';
  return '$bookName $start–$end';
}

File: apps/web/lib/plan_steps.ts
// 计划阅读 Step：最小跳转单元（如「马可 1–2」「约翰 3–6」）。

export interface PlanStep {
  id: string;
  label: string;
  bookId: string;
  chapterStart: number;
  chapterEnd: number;
}

export interface ReadingDayRow {
  day: number;
  book: string;
  book_name?: string;
  bookName?: string;
  chapter_start?: number;
  chapter_end?: number;
  chapterStart?: number;
  chapterEnd?: number;
  title?: string;
}

export interface GeneratedDayRow {
  day: number;
  title: string;
  refs: string[];
  date?: string;
}

function bookLabel(row: ReadingDayRow): string {
  return row.book_name ?? row.bookName ?? row.book;
}

function formatRange(bookName: string, start: number, end: number): string {
  if (start === end) return `${bookName} ${start}`;
  return `${bookName} ${start}–${end}`;
}

export function stepsForReadingRows(rows: ReadingDayRow[], day: number): PlanStep[] {
  return rows
    .filter((r) => r.day === day)
    .map((r) => {
      const cs = r.chapter_start ?? r.chapterStart ?? 1;
      const ce = r.chapter_end ?? r.chapterEnd ?? cs;
      const name = bookLabel(r);
      return {
        id: `${r.book}.${cs}-${ce}`,
        label: r.title?.trim() || formatRange(name, cs, ce),
        bookId: r.book.toUpperCase(),
        chapterStart: cs,
        chapterEnd: ce,
      };
    });
}

function parseRef(ref: string): { book: string; chapter: number } {
  const [book, ch] = ref.split('.');
  return { book: book.toUpperCase(), chapter: Number(ch) || 1 };
}

export function stepsFromRefs(refs: string[], titleHint?: string): PlanStep[] {
  if (!refs.length) return [];
  const parsed = refs.map(parseRef);
  const steps: PlanStep[] = [];
  let start = 0;
  for (let i = 1; i <= parsed.length; i++) {
    const prev = parsed[i - 1];
    const curr = parsed[i];
    const breakGroup =
      !curr || curr.book !== parsed[start].book || curr.chapter !== prev.chapter + 1;
    if (breakGroup) {
      const seg = parsed.slice(start, i);
      const book = seg[0].book;
      const cs = seg[0].chapter;
      const ce = seg[seg.length - 1].chapter;
      steps.push({
        id: `${book}.${cs}-${ce}`,
        label: formatRange(book, cs, ce),
        bookId: book,
        chapterStart: cs,
        chapterEnd: ce,
      });
      start = i;
    }
  }
  if (steps.length === 1 && titleHint) {
    steps[0] = { ...steps[0], label: titleHint };
  }
  return steps;
}

export function stepsForGeneratedDay(day: GeneratedDayRow): PlanStep[] {
  return stepsFromRefs(day.refs, day.title);
}

export function sessionProgress(steps: PlanStep[], stepsDone: string[]) {
  const done = steps.filter((s) => stepsDone.includes(s.id)).length;
  return { done, total: steps.length };
}

export function allStepsDone(steps: PlanStep[], stepsDone: string[]): boolean {
  return steps.length > 0 && steps.every((s) => stepsDone.includes(s.id));
}

export function stepForChapter(
  steps: PlanStep[],
  bookId: string,
  chapter: number,
): number {
  const bid = bookId.toUpperCase();
  return steps.findIndex(
    (s) =>
      s.bookId === bid && chapter >= s.chapterStart && chapter <= s.chapterEnd,
  );
}

export function isLastChapterOfStep(step: PlanStep, chapter: number): boolean {
  return chapter === step.chapterEnd;
}

export function nextIncompleteStep(
  steps: PlanStep[],
  stepsDone: string[],
): PlanStep | null {
  return steps.find((s) => !stepsDone.includes(s.id)) ?? null;
}

/** 当前位置在 Step 末章且仍有下一段未完成。 */
export function pendingNextStep(
  steps: PlanStep[],
  stepsDone: string[],
  bookId: string,
  chapter: number,
): PlanStep | null {
  const bid = bookId.toUpperCase();
  const step = steps.find(
    (s) => s.bookId === bid && chapter >= s.chapterStart && chapter <= s.chapterEnd,
  );
  if (!step || chapter !== step.chapterEnd) return null;
  const next = nextIncompleteStep(steps, stepsDone);
  if (!next || next.id === step.id) return null;
  return next;
}
